drivers/flash/flash_simulator: No real allocation just simulation

de-nordic opened the pull request:
The commit adds CONFIG_FLASH_SIMULATOR_NOALLOC Kconfig option
that will cause the flash simulator driver to allocate maximum
of write_block_size, while still respecting size and offset in DTS,
and pretending to do read/write/erase operations.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic committed:
drivers/flash/flash_simulator: No real allocation just simulation

The commit adds CONFIG_FLASH_SIMULATOR_NOALLOC Kconfig option
that will cause the flash simulator driver to allocate maximum
of write_block_size, while still respecting size and offset in DTS,
and pretending to do read/write/erase operations.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

nvlsianpu reviewed (changes requested):
need to add optional `memset` instead of `memcpy` in flash_smi_read(),
different otional memset size in unit_erase()
flash_mock_init(_ for native posix need also use smaller mock_flash size

carlescufi commented on drivers/flash/flash_simulator.c:
```diff
@@ -51,7 +51,12 @@
 #define MOCK_FLASH(addr) (mock_flash + (addr) - FLASH_SIMULATOR_BASE_OFFSET)
+#else
+#define MOCK_FLASH(addr) (mock_flash + ((addr) % FLASH_SIMULATOR_PROG_UNIT) - \
+			  FLASH_SIMULATOR_BASE_OFFSET)
```
This might go behind the starting address of `mock_flash`?

tejlmand commented on drivers/flash/Kconfig.simulator:
```diff
@@ -26,6 +26,14 @@ config FLASH_SIMULATOR_DOUBLE_WRITES
 	 Keep in mind that write operations can only pull bits to zero, regardless.
 
+config FLASH_SIMULATOR_NOALLOC
+	bool "No buffer allocated, reads, writes and erase are simulated"
```
as I read the code, those are not simulated, they are stubbed.

If I do multiple writes, I cannot readback later what I wrote from what I see.

This is a limitation that should be clearer, and because of this I would like a better term than `NOALLOC`, cause it's not just not allocating memory but still functioning in similar way, it's changing behavior also.

tejlmand reviewed (changes requested):
This is not simply a flash simulator anymore, instead it's a stub.

nvlsianpu commented on drivers/flash/Kconfig.simulator:
```diff
@@ -26,6 +26,14 @@ config FLASH_SIMULATOR_DOUBLE_WRITES
 	 If selected, writing to a non-erased program unit will succeed, otherwise, it will return an error.
 	 Keep in mind that write operations can only pull bits to zero, regardless.
 
+config FLASH_SIMULATOR_NOALLOC
```
depends on !NATIVE_POSIX

MarekPieta replied:
!ARCH_POSIX you mean ?

tejlmand replied:
As I mentioned in the review, this setting changes the simulator into a stub, so please change the name.

That way a user immediately understands that this will change behavior.

Especially, the name `NOALLOC` together with `will cause driver to allocate only write_block_size`.
actually means it will allocate memory, just less memory.

```suggestion
config FLASH_SIMULATOR_STUB
```

Laczen commented:
@de-nordic, what is the usecase for this change?

tejlmand commented:
also missing test of this new functionality.

Laczen commented:
> also missing test of this new functionality.

And if selected, this will break other tests.

de-nordic committed:
tests/drivers/flash_simulator_noalloc: No buffer flash simulator

The commit adds test for flash simulator with enabled
CONFIG_FLASH_SIMULATOR_NOALLOC option.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

de-nordic commented:
> 
> 
> @de-nordic, what is the usecase for this change?

You have two application that switch operations that share the same DTS with definition of large flash sim in RAM.
First app is tiny uses the RAM flash sim as a buffer, and uses and other flash devices.
The second app uses a lot of RAM, and can not afford the sim, but it can recognize valid data on flash devices by reading headers and comparing to erase values.
So the second application iterates over the same set of devices, as the first one knows, but the flash sim device will return 0xff headers, basically meaning that there is not interesting data, and will be skipped.

Laczen commented:
> > @de-nordic, what is the usecase for this change?
> 
> You have two application that switch operations that share the same DTS with definition of large flash sim in RAM.
> First app is tiny and uses the RAM flash sim as a buffer, and other flash devices.
> The second app uses a lot of RAM, and can not afford the sim, but it can recognize valid data on flash devices by reading headers and comparing to erase values.
> So the second application iterates over the same set of devices, as the first one knows, but the flash sim device will return 0xff headers, basically meaning that there is not interesting data, and will be skipped.

@de-nordic, is it not easier to just compile the second application with `CONFIG_FLASH_SIMULATOR=N` or is there a reason that they need to have the same set of devices ?

tejlmand reviewed (changes requested):
@de-nordic you still haven't addressed this comment: https://github.com/zephyrproject-rtos/zephyr/pull/38524#pullrequestreview-753876465
> This is not simply a flash simulator anymore, instead it's a stub.

carlescufi commented on drivers/flash/Kconfig.simulator:
```diff
@@ -26,6 +26,15 @@ config FLASH_SIMULATOR_DOUBLE_WRITES
 	 Keep in mind that write operations can only pull bits to zero, regardless.
 
+config FLASH_SIMULATOR_NOALLOC
+	bool "Reads, writes and erase are simulated within buffer of write_block_size size"
```
If we change the help to show `stub` then I think maybe we should consider renaming the Kconfig option to `FLASH_SIMULATOR_STUB`. No preference either way on my side.

tejlmand replied:
> If we change the help to show `stub` then I think maybe we should consider renaming the Kconfig option to `FLASH_SIMULATOR_STUB`. No preference either way on my side.

which is in my other comment here: https://github.com/zephyrproject-rtos/zephyr/pull/38524#discussion_r708591856 :wink:

Laczen commented:
@carlescufi, shouldn't this change be considered as an API change?

carlescufi commented:
> @carlescufi, shouldn't this change be considered as an API change?

I checked and I couldn't find any API changes if the new Kconfig option is not enabled (and it defaults to no), so AFAICT this is not a stable API change, but happy to be proven wrong.

Laczen commented:
> > @carlescufi, shouldn't this change be considered as an API change?
> 
> I checked and I couldn't find any API changes if the new Kconfig option is not enabled (and it defaults to no), so AFAICT this is not a stable API change, but happy to be proven wrong.

I didn't know API changes are only considered when default enabled, good to know.

What I was trying to ask is: Is (in general) adding a stub a change to the API ?

carlescufi commented:
> What I was trying to ask is: Is (in general) adding a stub a change to the API ?

No, at least not in this case. We typically define a change in API in this way:
https://docs.zephyrproject.org/latest/reference/api/api_lifecycle.html#introducing-incompatible-changes

>An incompatible change is defined as one that forces users to modify their existing code in order to maintain the current behavior of their application. The need for recompilation of applications (without changing the application itself) is not considered an incompatible change.

nvlsianpu commented:
@de-nordic Shouldn't we not proceed this anymore?

de-nordic commented:
> 
> 
> @de-nordic Shouldn't we not proceed this anymore?

No, I think we will be dropping this.

de-nordic commented:
implemented as other device.
